processor/src/main/java/org/mapstruct/ap/internal/model/source/SubClassMappingOptions.java

```diff
@@ -0,0 +1,135 @@
+
+/**
+ * Represents a sub class mapping as configured via {@code @SubClassMapping}.
+ *
+ * @author Ben Zegveld
+ */
+public class SubClassMappingOptions extends DelegatingOptions {
+
+    private TypeMirror sourceClass;
+    private TypeMirror targetClass;
+
+    public SubClassMappingOptions(TypeMirror sourceClass, TypeMirror targetClass, DelegatingOptions next) {
+        super( next );
+        this.sourceClass = sourceClass;
+        this.targetClass = targetClass;
+    }
+
+    @Override
+    public boolean hasAnnotation() {
+        return sourceClass != null && targetClass != null;
+    }
+
+    public static List<SubClassMappingOptions> getInstanceOn(ExecutableElement method, MapperOptions mapperOptions,
+                                                       FormattingMessager messager, TypeUtils typeUtils) {
+
+        SubClassMappingsGem subClassMappings = SubClassMappingsGem.instanceOn( method );
+        List<SubClassMappingGem> subClassMappingAnnotations;
+        if ( nonNull( subClassMappings ) ) {
+            subClassMappingAnnotations = subClassMappings.value().get();
+        }
+        else if ( nonNull( SubClassMappingGem.instanceOn( method ) ) ) {
+            subClassMappingAnnotations = Arrays.asList( SubClassMappingGem.instanceOn( method ) );
+        }
+        else {
+            return Collections.emptyList();
+        }
+
+        List<SubClassMappingOptions> subClassMappingOptions = new ArrayList<>();
+
+        for ( SubClassMappingGem subClassMapping : subClassMappingAnnotations ) {
+
+            if ( !isConsistent( subClassMapping, method, messager, typeUtils ) ) {
+                continue;
+            }
+
+            TypeMirror sourceSubClass = subClassMapping.sourceClass().getValue();
+            TypeMirror targetSubClass = subClassMapping.targetClass().getValue();
+
+            subClassMappingOptions
+                                  .add(
+                                      new SubClassMappingOptions(
+                                          sourceSubClass,
+                                          targetSubClass,
+                                          mapperOptions ) );
+        }
+        return subClassMappingOptions;
+    }
+
+    private static boolean isConsistent(SubClassMappingGem gem, ExecutableElement method, FormattingMessager messager,
+                                        TypeUtils typeUtils) {
+
+        if ( method.getParameters().isEmpty() || method.getReturnType().getKind() == TypeKind.VOID ) {
+            messager
+                    .printMessage(
+                        method,
+                        gem.mirror(),
+                        SUBCLASSMAPPING_METHOD_SIGNATURE_NOT_SUPPORTED );
+            return false;
+        }
+
+        TypeMirror sourceSubClass = gem.sourceClass().getValue();
+        TypeMirror targetSubClass = gem.targetClass().getValue();
+        TypeMirror sourceParentType = method.getParameters().get( 0 ).asType();
```

Comment:
What is the correct way to determine the sourceParentType? Currently assuming that the first argument of the method is the source that needs to be mapped to the target.

Reply:
See https://github.com/mapstruct/mapstruct/pull/2512/#discussion_r697980035

Reply:
Replaced this with checking if any parameter matches. That way it doesn't matter which parameter the source parameter is.

Reply:
I've left a comment about this in my last review. The current solution will also work for

```
@SubClassMapping(source = SubArg1, target = Target)
@SubClassMapping(source = SubArg2, target = Target)
Target map(SuperArg1 arg1, @Context SuperArg2 arg2);
```

However, it shouldn't. `@Context` arguments should not be treated as potential source.